Remove nom dependency

rfuest opened the pull request:
While I like `nom` I find it hard to implement error handling with `nom`. After fighting with `nom`s error types for a while I noticed that we actually use a very limited set of `nom` functions in this crate. So I decided to try to reimplement the necessary function and it turned out to only needed about 30 lines of code to implement all necessary functions.

The code is missing tests for the new errors and the parser functions, but it should be enough to decide if we want to stick with `nom` or use this approach.

jamwaffles commented:
This looks really clean so far, so while it's sad to see `nom` removed (I have quite the soft spot for it), this PR does present a good solution.

If you really wanted to, could you benchmark this PR against `master`? I'm curious to see if there's any performance improvement to be had.

rfuest commented:
> This looks really clean so far, so while it's sad to see `nom` removed (I have quite the soft spot for it), this PR does present a good solution.

I did notice your fondness for `nom` and I was hesitant to remove it, because I also like it. But the error handling code I tried to get the `nom` working was already longer than my minimal nom reimplementation.

> If you really wanted to, could you benchmark this PR against `master`? I'm curious to see if there's any performance improvement to be had.

I was also curious if there is any difference and added a bench for the BMP parser, but the result is inconclusive. Maybe the new version is a bit slower, but I get large run to run variations. I've added the benches as a separate commit if you want to run the benches yourself.

But parsing the BMP is only a small fraction of the overall drawing time and a small difference in this code won't matter as much in the total time. The pixel iterator doesn't use `nom`.

rfuest commented:
If you are OK with this code I would like to merge it and add additional tests and improved error handling as part of the next PR, which will try to fix the color type issues.

jamwaffles commented on src/header/mod.rs:
```diff
@@ -3,16 +3,15 @@
 //! Information gleaned from [wikipedia](https://en.wikipedia.org/wiki/BMP_file_format) and
 //! [this website](http://paulbourke.net/dataformats/bmp/)
 
+use embedded_graphics::prelude::*;
```
I have a feeling this was caused by a rust-analyzer autoimport, but either way I'd prefer not to use the prelude and/or wildcard imports.

Is it part of the standard order to have this, then `mod`, then `use crate`? That looks weird too but if it's part of the normal behaviour feel free to leave it.

rfuest replied:
> I have a feeling this was caused by a rust-analyzer autoimport, but either way I'd prefer not to use the prelude and/or wildcard imports.

It wasn't caused by rust-analyzer. I think we had previously discussed this and decided to not use the prelude inside e-g itself but use the prelude in examples and external creates. But if you have a good reason for not using the prelude it won't take much to convince me to change it.

> Is it part of the standard order to have this, then `mod`, then `use crate`? That looks weird too but if it's part of the normal behaviour feel free to leave it.

It is a weird and not part of any standard. Maybe like this?
```rust
use embedded_graphics::prelude::*;

use crate::{
    color_table::ColorTable,
    parser::{le_u16, le_u32, take, take_slice},
    ParseError,
};

mod dib_header;

use dib_header::DibHeader;
```
This form makes it easier to add a `pub` to the imports from child modules if necessary. But I'm open to other suggestions.

jamwaffles replied:
> decided to not use the prelude inside e-g itself but use the prelude in examples and external creates

Right, gotcha, let's leave it as is then.

> Maybe like this?

I like putting `mod` before any imports but that's just me. I'm happy to use your suggestion though. Either way let's not bikeshed this too much :)

```
mod dib_header;

use embedded_graphics::prelude::*;

use crate::{
    color_table::ColorTable,
    parser::{le_u16, le_u32, take, take_slice},
    ParseError,
};

use dib_header::DibHeader;
```

rfuest replied:
> I like putting `mod` before any imports but that's just me. I'm happy to use your suggestion though. Either way let's not bikeshed this too much :)

It is bikeshedding, but I would prefer to use one common style across all e-g crates. I've used my last suggestion for now and adjusted all other imports in the crate to match.

My reason for putting the mods in between the imports is to divide imports from other crates from child module imports. Otherwise it's not as clear that `use dib_header::DibHeader` isn't importing something from a `dib_header` crate. If it is preceded by `mod dib_header` it should be obvious that it's an import from a child module.

jamwaffles reviewed (changes requested):
One nit/question, otherwise lgtm. Nice use of const generics too.

We never had an MSRV for tinybmp IIRC, but it might be worth adding a changelog entry to this PR to state the MSRV as the one which introduced min-const-generics (or 1.57 in line with e-g).

rfuest commented:
> We never had an MSRV for tinybmp IIRC, but it might be worth adding a changelog entry to this PR to state the MSRV as the one which introduced min-const-generics (or 1.57 in line with e-g).

I would suggest to use 1.57, otherwise we need to maintain another docker image to test the code with the MSRV.

IMO this change isn't directly related to this PR. The MSRV also needs to be added to the README/docs, which need to be updated when the API is changed. Can we postpone this until the breaking API changes are added? It would also be a good time to change to intra-doc links and the new README tooling from the e-g repo.

jamwaffles reviewed (approved):
Yep, that's fine. I agree it makes sense to migrate everything in its own PR.

rfuest commented:
Thanks for the review!
